File: src/js/src/data/Port.ts
import type { RawPort } from './types';
import type { Voyage } from './Voyage';

/**
 * A port (e.g. "AAL" = AALBORG). Navigable to the voyages that start or end
 * here. Note: in the current sample data most voyages have null ports, so
 * these links are often empty — the relationship is modelled regardless.
 */
export class Port {
  readonly code: string;
  readonly description: string;

  private readonly _voyagesFrom: Voyage[] = [];
  private readonly _voyagesTo: Voyage[] = [];

  constructor(raw: RawPort) {
    this.code = (raw.code ?? '').trim();
    this.description = (raw.description ?? '').trim();
  }

  /** Voyages departing from this port. */
  get voyagesFrom(): readonly Voyage[] {
    return this._voyagesFrom;
  }

  /** Voyages terminating at this port. */
  get voyagesTo(): readonly Voyage[] {
    return this._voyagesTo;
  }

  /** @internal */
  _addVoyageFrom(voyage: Voyage): void {
    this._voyagesFrom.push(voyage);
  }

  /** @internal */
  _addVoyageTo(voyage: Voyage): void {
    this._voyagesTo.push(voyage);
  }
}

File: src/dotnet/ApiSdk/Data/Port.cs
namespace ApiSdk.Data;

/// <summary>
/// A port (e.g. "AAL"). Navigable to the voyages that start or end here. Links
/// are often empty in the sample data (voyages have null ports); the
/// relationship is modelled regardless.
/// </summary>
public sealed class Port
{
    private readonly List<Voyage> _voyagesFrom = new();
    private readonly List<Voyage> _voyagesTo = new();

    public string Code { get; }
    public string Description { get; }

    internal Port(string code, string description)
    {
        Code = code;
        Description = description;
    }

    public IReadOnlyList<Voyage> VoyagesFrom => _voyagesFrom;
    public IReadOnlyList<Voyage> VoyagesTo => _voyagesTo;

    internal void AddVoyageFrom(Voyage voyage) => _voyagesFrom.Add(voyage);
    internal void AddVoyageTo(Voyage voyage) => _voyagesTo.Add(voyage);
}
